Libraries/MLXVLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,743 @@
+        default: throw VLMError.singleImageAllowed
+        }
+
+        // this doesn't have a chat template so just use the last message.
+        var prompt = input.prompt.asMessages().last?["content"] ?? ""
+
+        // based on transformers/processing_paligemma
+        let count = input.images.count * config.imageSequenceLength
+        prompt =
+            Array(repeating: "<image>", count: count).joined() + (tokenizer.bosToken ?? "") + prompt
+            + "\n"
+
+        let promptTokens = try tokenizer.encode(text: prompt)
+        let promptArray = MLXArray(promptTokens).expandedDimensions(axis: 0)
+        let mask = ones(like: promptArray)
+
+        let pixels = try prepare(image: input.images[0].asCIImage(), processing: input.processing)
+
+        return LMInput(text: .init(tokens: promptArray, mask: mask), image: .init(pixels: pixels))
+    }
+
+}
+
+// MARK: - Model
+
+private class PaliGemmaMultiModalProjector: Module, UnaryLayer {
+
+    @ModuleInfo var linear: Linear
+
+    public init(_ config: PaliGemmaConfiguration.VisionConfiguration) {
+        self.linear = Linear(config.hiddenSize, config.projectionDimensions, bias: true)
+    }
+
+    public func callAsFunction(_ x: MLXArray) -> MLXArray {
+        linear(x)
+    }
+}
+
+/// PaliGemma VLM
+///
+/// This is typically created by ``VLMModelFactory``.
+public class PaliGemma: Module, VLMModel, KVCacheDimensionProvider {
+
+    @ModuleInfo(key: "vision_tower") private var visionModel: Vision.VisionModel
+    @ModuleInfo(key: "language_model") private var languageModel: Language.LanguageModel
+    @ModuleInfo(key: "multi_modal_projector") private var multiModalProjector:
+        PaliGemmaMultiModalProjector
+
+    public let config: PaliGemmaConfiguration
+
+    public var vocabularySize: Int { config.vocabularySize }
+    public var kvHeads: [Int] { languageModel.kvHeads }
+    public var headDim: MLX.IntOrPair { languageModel.headDim }
+
+    public func loraLinearLayers() -> MLXLMCommon.LoRALinearLayers {
+        languageModel.model.layers.map { ($0.attention, ["q_proj", "v_proj"]) }
+    }
+
+    public init(_ config: PaliGemmaConfiguration) {
+        self.config = config
+        self._visionModel.wrappedValue = Vision.VisionModel(config.visionConfiguration)
+        self._languageModel.wrappedValue = Language.LanguageModel(config.textConfiguration)
+        self._multiModalProjector.wrappedValue = PaliGemmaMultiModalProjector(
+            config.visionConfiguration)
+    }
+
+    private func inputEmbeddings(inputIds: MLXArray, pixelValues: MLXArray?, mask: MLXArray) -> (
+        MLXArray, MLXArray
+    ) {
+        guard let pixelValues else {
+            return (inputIds, mask)
+        }
+
+        let inputEmbedding = languageModel.model.embedTokens(inputIds)
+        let (hiddenState, _, _) = self.visionModel(
+            pixelValues.transposed(0, 2, 3, 1).asType(inputEmbedding.dtype),
+            outputHiddenStates: true
+        )
```

Comment:
We have to be pretty careful with data types in these models cause it's really easy to upcast to fp32 by accident and that can slow things down a lot or use a lot more memory (or both).

One thing I recommend doing is if you have a test suite that runs the models, making sure the output type is the same as the input type.

Here you cast the `pixelValues` to the embedding type which is good. But below you cast the output back to the `pixelValues` type which I'm not sure about.. I would just keep those in the same model type.

Reply:
Good spot on that!

```
inputEmbedding float16, hiddenState float32, pixelValues float32
```